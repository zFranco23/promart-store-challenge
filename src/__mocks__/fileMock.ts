export const mockProducts = [
  {
    id: 1,
    title: 'title1',
    price: 5,
    category: 'c1',
    description: 'd1',
    image: 'image',
    rating: {
      count: 5,
      rate: 10,
    },
  },
  {
    id: 2,
    title: 'title2',
    price: 10,
    category: 'c2',
    description: 'd2',
    image: 'image',
    rating: {
      count: 6,
      rate: 19,
    },
  },
];
