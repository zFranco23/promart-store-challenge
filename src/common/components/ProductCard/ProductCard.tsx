import styled from "styled-components";
import { useNavigate } from "react-router-dom";

import { Product } from "../../../entities";
import { fadeIn } from "../../animations/fadeIn";
import { ShoppingCartItem } from "../../../entities/shopping-cart";
import { useAppDispatch } from "../../../hooks/store";

import { actions as scActions } from "../../../modules/shopping-cart/duck";

const ProductWrap = styled.div`
  position: relative;
  height: 100%;
  max-height: 100%;
  display: flex;
  flex-direction: column;
  // background-color: #ebeeef;
  border: 2px solid #eaeaea;
  border-radius: 15px;
  overflow: hidden;
  box-shadow: 0px 0px 20px rgba(0, 0, 0, 0.1);
  transition: all 0.3s ease-in-out;
  animation: ${fadeIn} 1.2s cubic-bezier(0.39, 0.575, 0.565, 1) both;

  &:hover {
    border: 2px solid #ff6e00;
    transform: scale(1.02);
    box-shadow: 0px 0px 30px rgba(0, 0, 0, 0.2);
  }

  &:hover img {
    filter: grayscale(0);
  }

  &:hover button {
    display: flex;
  }
`;

const ProductImage = styled.img`
  display: block;
  width: 100%;
  height: 200px;
  object-fit: contain;
  transition: all 0.3s ease-in-out;
  filter: grayscale(0.3);
`;

const ProductDetails = styled.div`
  padding: 1rem;
  height: 100%;
  max-height: 100%;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
`;

const ProductName = styled.h3`
  margin-bottom: 1.5rem;
  font-size: 1.8rem;
  font-weight: 700;
  color: #333;
`;

const ProductPrice = styled.p`
  margin: 0;
  font-size: 1.6rem;
  font-weight: 700;
  color: #333;
`;

const ProductDescription = styled.p`
  margin: 0 0 0.5rem;
  font-size: 1rem;
  color: #666;
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  line-clamp: 2;
  -webkit-box-orient: vertical;
`;

const AddToCartButton = styled.button`
  z-index: 2;
  cursor: pointer;
  position: absolute;
  right: 8px;
  top: 8px;
  width: 3.5rem;
  height: 3.5rem;
  border-radius: 50%;
  background-color: #ff6e00;
  color: #fff;
  font-weight: 700;
  display: flex;
  align-items: center;
  justify-content: center;
  border: none;
  animation: ${fadeIn} 0.3s cubic-bezier(0.39, 0.575, 0.565, 1) both;

  @media (min-width: 768px) {
    display: none;
  }
`;

type Props = {
  product: Product;
};
const ProductCard = (props: Props) => {
  const { product } = props;

  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { image, title, price, description } = product;

  const onAddToCart = () => {
    const scItem: ShoppingCartItem = {
      ...product,
      quantity: 1,
    };
    dispatch(scActions.addItemToCart(scItem));
    navigate("/cart");
  };

  return (
    <ProductWrap>
      <AddToCartButton onClick={onAddToCart}>
        <i className="material-icons">add</i>
      </AddToCartButton>
      <ProductImage src={image} alt={title} />
      <ProductDetails>
        <ProductName>{title}</ProductName>
        <div className="bg-neutral10 rounded-2xl p-2">
          <ProductDescription>{description}</ProductDescription>
          <ProductPrice>S/. {price}</ProductPrice>
        </div>
      </ProductDetails>
    </ProductWrap>
  );
};

export default ProductCard;
