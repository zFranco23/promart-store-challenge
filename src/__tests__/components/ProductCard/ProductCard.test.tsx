import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { shallow } from 'enzyme';
import { mockProducts } from '../../../__mocks__/fileMock';
// import { ShoppingCartItem } from '../../../entities';

import ProductCard from '../../../common/components/ProductCard/ProductCard';
import { BrowserRouter, Route, Routes } from 'react-router-dom';

const mockStore = configureStore();

describe('Product Card tests', () => {
  let store: ReturnType<typeof mockStore>;
  let wrapper: React.ReactElement;
  beforeEach(() => {
    store = mockStore({
      currency: { code: 'Sol Peruano', symbol: 'S/.' },
      shoppingCart: {
        items: [mockProducts[0]],
      },
    });
    wrapper = (
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path='*' element={<ProductCard product={mockProducts[0]} />} />
          </Routes>
        </BrowserRouter>
      </Provider>
    );
  });

  test('Product card render correctly', () => {
    expect(shallow(wrapper)).toMatchSnapshot();
  });

  // test('Should add product to cart ', () => {
  //   const component = shallow(wrapper);
  //   const productCard = component.find('ProductCard');

  //   console.log(component.html());

  //   const addButton = productCard.find('button');
  //   console.log(addButton);
  //   addButton.simulate('click');
  //   // expect(component.instance).toBe(1);
  // });
});
