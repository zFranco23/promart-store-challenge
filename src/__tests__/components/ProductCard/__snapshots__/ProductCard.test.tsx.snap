// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card tests Product card render correctly 1`] = `
<ContextProvider
  value={
    {
      "getServerState": undefined,
      "store": {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <BrowserRouter>
    <Routes>
      <Route
        element={
          <ProductCard
            product={
              {
                "category": "c1",
                "description": "d1",
                "id": 1,
                "image": "image",
                "price": 5,
                "rating": {
                  "count": 5,
                  "rate": 10,
                },
                "title": "title1",
              }
            }
          />
        }
        path="*"
      />
    </Routes>
  </BrowserRouter>
</ContextProvider>
`;
